### PR TITLE
feat: Initial commit for EnterpriseDB Module

### PR DESCRIPTION
**Pull request by imprateeksh**
### Description

Initial changes done for ICD EDB Module.

#### Changes that affect the core Terraform module or submodules
- [ ] Bug fix
- [x] New feature
- [ ] Dependency update

#### Changes that don't affect the core Terraform module or submodules
- [x] Examples or tests (addition or updates of examples or tests)
- [x] Documentation update
- [ ] CI-related update (pipeline, etc.)
- [ ] Other

#### Release required?
Identify the type of release. For information about the changes in a semantic versioning release, see [Release versioning](https://terraform-ibm-modules.github.io/documentation/#/versioning).

- [ ] No release
- [ ] Patch release (`x.x.X`)
- [ ] Minor release (`x.X.x`)
- [x] Major release (`X.x.x`)

##### Release notes content

Initial ICD Enterprise DB module release.

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### Merge actions for mergers

- Use a relevant [conventional commit](https://www.conventionalcommits.org/) message that is based on the PR contents and any release notes provided by the PR author. The commit message determines whether a new version of the module is needed, and if so, which semver increment to use (major, minor, or patch).
- Merge by using "Squash and merge".

### CODE REVIEW

**Review by ocofaigh — CHANGES_REQUESTED** [review]
Can you change back the ci.yml and release.yml so it runs new pipeline (same as module template)? That means that a comment `/run pipeline` will then be nedded to trigger pipeline.

**Comment by imprateeksh** [comment]
> Can you change back the ci.yml and release.yml so it runs new pipeline (same as module template)? That means that a comment `/run pipeline` will then be nedded to trigger pipeline.

Sure. Reverted the changes of`ci.yml` and `release.yml`

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -8,9 +8,10 @@ on:
 
 jobs:
   call-terraform-ci-pipeline:
-    uses: terraform-ibm-modules/common-pipeline-assets/.github/workflows/common-terraform-module-ci-v2.yml@v1.11.5
+    uses: terraform-ibm-modules/common-pipeline-assets/.github/workflows/common-terraform-module-ci-v2.yml@v1.12.2
```
Comment (ocofaigh) [inline .github/workflows/ci.yml:11]: bump to v1.14.0 (has latest fixes)

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -8,5 +8,5 @@ on:
 
 jobs:
   call-terraform-release-pipeline:
-    uses: terraform-ibm-modules/common-pipeline-assets/.github/workflows/common-release.yml@v1.11.5
+    uses: terraform-ibm-modules/common-pipeline-assets/.github/workflows/common-release.yml@v1.12.2
```
Comment (ocofaigh) [inline .github/workflows/release.yml:11]: bump to v1.14.0 (has latest fixes)

**Review by ocofaigh — COMMENTED** [review]
Please ensure the changes that went into https://github.com/terraform-ibm-modules/terraform-ibm-icd-postgresql/pull/204 today are also reflected in this PR

File: .github/workflows/ci.yml (hunk 1 above)
Comment (ocofaigh) [reply .github/workflows/ci.yml:11]: never mind, I merged it into main branch

File: .github/workflows/release.yml (hunk 2 above)
Comment (ocofaigh) [reply .github/workflows/release.yml:11]: never mind, I merged it into main branch

**Comment by ocofaigh** [comment]
@imprateeksh Once you resolve conflicts, the new pipeline version will allow this PR to be triggered by comment

**Review by ocofaigh — CHANGES_REQUESTED** [review]
`admin_pass` and `users` should be added to the fscloud profile module too

File: .github/workflows/release.yml (hunk 2 above)
Comment (imprateeksh) [reply .github/workflows/release.yml:11]: sure.

File: .github/workflows/ci.yml
Comment (rajatagarwal-ibm) [inline .github/workflows/ci.yml:1]: Need to add controls on push and pull requests:

```
# Controls when the workflow will run
on:
  # Triggers the workflow on push or pull request events but only for the main branch
  push:
    branches: [main]
  pull_request:
    branches: [main]
    types: [opened, synchronize, reopened, ready_for_review]
```

File: cra-tf-validate-ignore-rules.json [hunk 3]
```diff
@@ -1,3 +1,16 @@
 {
-    "scc_rules": []
-}
+    "scc_rules": [
+      {
+        "scc_rule_id": "rule-9b2d8054-bc93-44fd-901b-91f677287e84",
+        "description": "Check whether Databases for PostgreSQL network access is restricted to a specific IP range",
```
Comment (rajatagarwal-ibm) [inline cra-tf-validate-ignore-rules.json:5]: Have you run the SCC scan against this module? Based on the results you may need to update the rule id and description. If you need help running SCC scan you can reach out to me.

File: outputs.tf [hunk 4]
```diff
@@ -2,9 +2,49 @@
...
+}
+
+output "cbr_rule_ids" {
+  description = "CBR rule ids created to restrict Enterprise DB"
+  value       = module.cbr_rule[*].rule_id
+}
+
+output "hostname" {
+  description = "Database hostname. Only contains value when var.service_credential_names or var.users are set."
+  value       = length(var.service_credential_names) > 0 ? nonsensitive(ibm_resource_key.service_credentials[keys(var.service_credential_names)[0]].credentials["connection.postgres.hosts.0.hostname"]) : length(var.users) > 0 ? nonsensitive(flatten(data.ibm_database_connection.database_connection[0].postgres[0].hosts[0].hostname)) : null
```
Comment (rajatagarwal-ibm) [inline outputs.tf:44]: These `connection strings` should point to the Enterprise DB?

File: .github/workflows/ci.yml
Comment (rajatagarwal-ibm) [reply .github/workflows/ci.yml:1]: Or maybe not, as we now run workflows based on the comments. @ocofaigh can you please confirm?

File: variables.tf [hunk 5]
```diff
@@ -2,10 +2,284 @@
...
+
+variable "pitr_id" {
+  type        = string
+  description = "(Optional) The ID of the source deployment EDB instance that you want to recover back to. The EDB instance is expected to be in an up and in running state."
+  default     = null
+}
+
+variable "pitr_time" {
+  type        = string
+  description = "(Optional) The timestamp in UTC format (%Y-%m-%dT%H:%M:%SZ) that you want to restore to. To retrieve the timestamp, run the command (ibmcloud cdb postgresql earliest-pitr-timestamp <deployment name or CRN>). For more info on Point-in-time Recovery, see https://cloud.ibm.com/docs/databases-for-enterprisedb?topic=databases-for-enterprisedb-pitr"
```
[the remark this answers is not in the document]
Comment (imprateeksh) [reply variables.tf:283]: This looks correct as EDB internally uses postgres. 
The command can be checked [here](https://cloud.ibm.com/docs/databases-for-enterprisedb?topic=databases-for-enterprisedb-pitr&interface=ui)
`ibmcloud cdb postgresql earliest-pitr-timestamp <deployment name or CRN>`

File: variables.tf [hunk 6]
```diff
@@ -2,10 +2,284 @@
...
+}
+
+variable "remote_leader_crn" {
+  type        = string
+  description = "A CRN of the leader database to make the replica(read-only) deployment. The leader database is created by a database deployment with the same service ID. A read-only replica is set up to replicate all of your data from the leader deployment to the replica deployment by using asynchronous replication. For more information, see https://cloud.ibm.com/docs/databases-for-enterprisedb?topic=databases-for-enterprisedb-read-only-replicas"
+  default     = null
+}
+
+variable "edb_version" {
+  description = "Version of the PostgreSQL instance to provision. If no value is passed, the current preferred version of IBM Cloud Databases is used."
```
[the remark this answers is not in the document]
Comment (imprateeksh) [reply variables.tf:28]: Thanks, made the changes.

[the remark this answers is not in the document]
Comment (imprateeksh) [reply variables.tf:28]: However, there is not much information about EDB versions. Only version 12 is seen when try to provision EDB from IBM Cloud

File: outputs.tf (hunk 4 above)
Comment (imprateeksh) [reply outputs.tf:44]: The EDB internally uses postgres hence the connection strings are correct. 
I have checked these after provisioning the EDB instance. Another instance provision is in progress, will attach the .json file too.

File: examples/complete/main.tf [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+# VPE
+##############################################################################
+
+resource "ibm_is_security_group" "sg1" {
+  name = "${var.prefix}-sg1"
+  vpc  = ibm_is_vpc.example_vpc.id
+}
+
+resource "ibm_is_virtual_endpoint_gateway" "pgvpe" {
+  name = "${var.prefix}-vpe-to-pg"
```
[the remark this answers is not in the document]
Comment (imprateeksh) [reply examples/complete/main.tf:119]: Thanks. I have made the changes.

**Comment by imprateeksh** [comment]
> 
Yes @rajatagarwal-ibm , I am working on a different branch to add remaining examples. 
However in case of basic example, I am not able to get the correct `member_memory_mb` and `member_disk_mb` values required for read_only_replica
In case of postgres, i can see these values as `2304` and `15360` but in the [Enterprise DB documentation](https://cloud.ibm.com/docs/databases-for-enterprisedb?topic=databases-for-enterprisedb-read-only-replicas&interface=ui), I am not able to get this value nor the right incremental value. Can you please help me here?

I tried with values as 3072 for RAM and 61440 for disk but it failed.

File: cra-tf-validate-ignore-rules.json (hunk 3 above)
Comment (imprateeksh) [reply cra-tf-validate-ignore-rules.json:5]: Thanks @rajatagarwal-ibm I will reach out to you.

File: outputs.tf (hunk 4 above)
Comment (imprateeksh) [reply outputs.tf:44]: This is from the credentials json created. Connection string `credentials["connection.postgres.hosts.0.hostname"]` is valid here. 
image

File: outputs.tf [hunk 8]
```diff
@@ -2,9 +2,49 @@
...
+}
+
+output "port" {
+  description = "Database port. Only contains value when var.service_credential_names or var.users are set."
+  value       = length(var.service_credential_names) > 0 ? nonsensitive(ibm_resource_key.service_credentials[keys(var.service_credential_names)[0]].credentials["connection.postgres.hosts.0.port"]) : length(var.users) > 0 ? nonsensitive(flatten(data.ibm_database_connection.database_connection[0].postgres[0].hosts[0].port)) : null
```
[the remark this answers is not in the document]
Comment (imprateeksh) [reply outputs.tf:49]: As shown in above image, even for port, the same connection string i.e. `connection.postgres.hosts.0.port` holds good.

File: main.tf [hunk 9]
```diff
@@ -1,3 +1,215 @@
...
+}
+
+locals {
+  # used for output only
+  service_credentials_json = length(var.service_credential_names) > 0 ? {
+    for service_credential in ibm_resource_key.service_credentials :
+    service_credential["name"] => service_credential["credentials_json"]
+  } : null
+
+  service_credentials_object = length(var.service_credential_names) > 0 ? {
```
[the remark this answers is not in the document]
Comment (imprateeksh) [reply main.tf:195]: `service_credentials` when created have reference of postgres.Only crn has info about enterprise db.
can be viewed [here](https://github.ibm.com/PRATEEK-SHARMA13/self-notes/blob/main/edb-svc-credentials)
I will be deleting this instance in next few minutes.

File: .github/workflows/ci.yml
Comment (ocofaigh) [reply .github/workflows/ci.yml:1]: no longer needed. Copy what is currently in https://github.com/terraform-ibm-modules/terraform-ibm-module-template

File: tests/pr_test.go [hunk 10]
```diff
@@ -2,43 +2,97 @@
...
 
 	"github.com/stretchr/testify/assert"
+	"github.com/terraform-ibm-modules/ibmcloud-terratest-wrapper/common"
 	"github.com/terraform-ibm-modules/ibmcloud-terratest-wrapper/testhelper"
 )
 
 // Use existing resource group
-const resourceGroup = "geretain-test-resources"
-const defaultExampleTerraformDir = "examples/default"
+const resourceGroup = "geretain-test-enterprisedb"
```
Comment (ocofaigh) [inline tests/pr_test.go:17]: Did you manually add this to our account? Please do not manually add permanent resource. They need to be added in https://github.ibm.com/GoldenEye/ge-dev-account-management/blob/master/main.tf and deployed with terraform

Comment (imprateeksh) [reply tests/pr_test.go:17]: You are right. I have not added yet. I have raised draft PR to create the resource group and EDB instances by this name. I guess I have to raise 2 PRs here. One for resource group name creation and another for provisioning EDB using main branch once changes are merged.

Comment (imprateeksh) [reply tests/pr_test.go]: I have added the default resource as `geretain-test-resources` and will be modifying this later once workspace issue is fixed.

File: README.md [hunk 11]
```diff
@@ -1,168 +1,121 @@
...
-<!-- Remove the content in this previous H2 heading -->
-## Reference architectures
-
-<!--
-Add links to any reference architectures for this module.
-(Usually in the `/reference-architectures` directory.)
-See "Reference architecture" in Authoring Guidelines in the public documentation at
-https://terraform-ibm-modules.github.io/documentation/#/implementation-guidelines?id=reference-architecture
--->
+[![Stable (With quality checks)](https://img.shields.io/badge/Status-Stable%20(With%20quality%20checks)-green?style=plastic)](https://terraform-ibm-modules.github.io/documentation/#/badge-status)
```
[the remark this answers is not in the document]
Comment (imprateeksh) [reply README.md:3]: Sure. I have added the mentioned badges.

**Review by ocofaigh — APPROVED** [review]

File: tests/pr_test.go (hunk 10 above)
Comment (imprateeksh) [reply tests/pr_test.go]: Sure. I will update this. Currently some issue is seen in workspace while provisioning new Resource group, hence using `geretain-test-resources` for now.

**Pull request merged**